### PR TITLE
Contingency decomposition with ExaTron

### PR DESCRIPTION
**Pull request by AnirudhSubramanyam**
* Contingency decomposition now supported with ExaTron (only when contingencies are linked via corrective control)
* Sequential tests have also been updated to include ExaTron
* ModelInfo now includes a parameter `corr_scale` that controls how tightly corrective control is enforced

### CODE REVIEW

File: src/Evaluators/ProxALEvalutor.jl [hunk 1]
```diff
@@ -278,7 +294,7 @@ function optimize!(nlp::ProxALEvaluator; print_timings=false)
     function proximal_parameter_update()
         elapsed_t = @elapsed begin
             if algparams.updateτ && runinfo.iter > 1
-                maxθ = algparams.decompCtgs  ? max(algparams.θ_t, algparams.θ_c) : algparams.θ_t
+                maxθ = (algparams.decompCtgs && modelinfo.num_ctgs > 0)  ? max(algparams.θ_t, algparams.θ_c) : algparams.θ_t
```
Comment (frapac) [inline src/Evaluators/ProxALEvalutor.jl:297]: How about defining a small function for this test? Like
```julia
has_slack(modelinfo, algparams) = (algparams.decompCtgs && modelinfo.num_ctgs > 0)
```

File: src/backends.jl [hunk 2]
```diff
@@ -566,13 +567,13 @@ function TronBlockBackend(
...
     use_gpu = (algparams.device == CUDADevice)
-    iterlim = 800
     trondata = ExaTron.OPFData(opfdata)
     env = ExaTron.ProxALAdmmEnv(
         trondata, use_gpu, t, T, algparams.tron_rho_pq, algparams.tron_rho_pa;
+        allow_infeas=modelinfo.allow_constr_infeas, rho_sigma=modelinfo.weight_constr_infeas,
         verbose=algparams.verbose_inner, use_twolevel=true, outer_eps=algparams.tron_outer_eps,
     )
-    return TronBlockBackend(env, blk, k, t, T, opfdata, modelinfo, iterlim, scale, modelinfo.obj_scale)
+    return TronBlockBackend(env, blk, k, t, T, opfdata, modelinfo, scale, modelinfo.obj_scale)
```
Comment (frapac) [inline src/backends.jl:576]: We don't need `iterlim` anymore?

File: src/backends.jl [hunk 3]
```diff
@@ -642,6 +649,70 @@ function set_objective!(block::TronBlockBackend, algparams::AlgParams,
...
+    # (we are defining the penalties manually)
+    examodel.model.gen_mod.rho = 0.0
+    examodel.model.gen_mod.tau = 0.0
+    # Reset Q_ref and c_ref
+    examodel.model.gen_mod.Q_ref .= 0.0
+    examodel.model.gen_mod.c_ref .= 0.0
+
+    # proximal terms
+    pg_ref = primal.Pg[:, k, t]
+    examodel.model.gen_mod.Q_ref[1:4:4*ngen] .+= algparams.τ/σ
```
Comment (frapac) [inline src/backends.jl:664]: I would suggest to define the indexes in a separate variable, for clarity purpose: 
```julia
index_generators_Q = 1:4:4*ngen
index_generators_c = 1:2:2*ngen
index_slacks_Q = ...
```

File: src/backends.jl [hunk 4]
```diff
@@ -642,6 +649,70 @@ function set_objective!(block::TronBlockBackend, algparams::AlgParams,
...
+
+            examodel.model.gen_mod.Q_ref[1:4:4*ngen] .+= (K - 1)*algparams.ρ_c/σ
+            examodel.model.gen_mod.c_ref[1:2:2*ngen] .+= -(pgc*algparams.ρ_c/σ) .+ λc
+        end
+
+        # contingencies
+        if k > 1
+            λf = dual.ctgs[:, k, t] ./ σ
+            pgf = primal.Pg[:, 1, t] .+ primal.Zk[:, k, t] .- scen_agc
+            examodel.model.gen_mod.Q_ref .+= algparams.ρ_c*repeat([1., -1., -1., 1.], ngen)/σ
```
Comment (frapac) [inline src/backends.jl:711]: Nice! Maybe we could rewrite everything with broadcast?
```julia
examodel.model.gen_mod.Q_ref .+= algparams.ρ_c .* repeat([1., -1., -1., 1.], ngen) ./ σ
```

**Review by frapac — COMMENTED** [review]
Nice work! I only have a few minor comments, otherwise I think it's ready to merge

File: src/backends.jl (hunk 4 above)
Comment (AnirudhSubramanyam) [reply src/backends.jl:711]: Done!

File: src/backends.jl (hunk 3 above)
Comment (AnirudhSubramanyam) [reply src/backends.jl:664]: Nice idea! Done.

File: src/backends.jl (hunk 2 above)
Comment (AnirudhSubramanyam) [reply src/backends.jl:576]: It appears that the relevant parameters are `algparams.tron_outer_iterlim` and `algparams.tron_inner_iterlim`

File: src/Evaluators/ProxALEvalutor.jl (hunk 1 above)
Comment (AnirudhSubramanyam) [reply src/Evaluators/ProxALEvalutor.jl:297]: Nice. Done.

File: src/backends.jl (hunk 2 above)
Comment (frapac) [reply src/backends.jl:582]: Good catch, indeed!

**Pull request merged**